### ISSUE TITLE
No JVM Found For Dockerfile.rhel7 When Built in OpenShift 3.3.1.5

### ISSUE THREAD

**Issue by sherl0cks** [sonatype/docker-rhel-nexus#17]
**Error** 
From deployer pod logs in OCP 3.3.1.5 
```
No suitable Java Virtual Machine could be found on your system.
The version of the JVM must be at least 1.8 and at most 1.8.
Please define INSTALL4J_JAVA_HOME to point to a suitable JVM.
```

**Steps to Reproduce** 

1. To make life easier, I refactored `Dockerfile` to extend the rhel image instead of centos so I wouldn't need to set the dockerfilepath in OpenShift. I pushed that to https://github.com/sherl0cks/docker-rhel-nexus/tree/no-jvm
2. `oc new-app  https://github.com/sherl0cks/docker-rhel-nexus.git#no-jvm`
3. Build passes
4. Deploy fails

**Comment by jorgemoralespou** [comment]
Building a RHEL  image that needs to install software should only work on a subscribed host. Is your openshift host subscribed?

cc/ @gordillo-ramon-redhat can you look into this?

**Comment by sherl0cks** [comment]
> Is your openshift host subscribed?

yes. 

The build succeeds just fine. It's the deploy that fails.

**Comment by sherl0cks** [comment]
@jorgemoralespou @gordillo-ramon-redhat 

FWIW - I just tested this on RHEL 7.2 outside of OpenShift and I can build and run the container image fine. Thus, this points to an issue with user permissions when running the image inside of kube.

I did have one issue (java.io.FileNotFoundException: /opt/sonatype/nexus/etc/org.apache.karaf.command.acl.bundle.cfg (Permission denied)) but this issue also found on https://github.com/sonatype/docker-nexus3 and doesn't seem to affect behavior.

**Comment by ghost** [comment]
I fixed that by setting the user prior to decompress the zip file. The
issue is regarding decompressing as root and running as othe user.

**Comment by sherl0cks** [comment]
@gordillo-ramon-redhat do you have a gist or a git ref you can point me to? I've tried the above approach and I still have the same error in OCP upon deployment

**Comment by sherl0cks** [comment]
@gordillo-ramon-redhat @jorgemoralespou bump. We'd really like to deploy a nexus on a RHEL base image. Do let me know if you have a working example that you can point me to. I'm unable to make changes that would allow this to work on OpenShift.

**Comment by ghost** [comment]
I have a PR for this fix.

In the meantime, you can check http://github.com/gordillo-ramon-redhat/docker-rhel-nexus.
When using Openshift Template, please double check to update the git url that is used to build the image accordingly. (sonatype -> gordillo-ramon-redhat until PR is merged).

**Comment by whyjustin** [comment]
@sherl0cks

I made some updates to the init scripts (using @gordillo-ramon-redhat suggestion for anyuid ssc) as well as many other places. Everything works for me now, although I kept Nexus home and Java owned by root (the mutable data and configuration are owned by the application user. Take a look and let me know if this solves your problem and I will close out #20

**Comment by sherl0cks** [comment]
@whyjustin Are the changes you refer to merge into master in this repo (as I see you made a commit around the same time as your comment). All tests done with master of this repo. In OpenShift 3.3.1.5, I can get CentOS working with the anyuid ssc command, but RHEL7 still fails. What am I missing here? FWIW, I'm not running the init.sh and I'm not deploying this via the template, because that isn't my use case. I simply want to run this via `oc new-app` like any other app.

**CentOS Dockerfile**

1. `oc new-app https://github.com/sonatype/docker-rhel-nexus` This creates a build and deploy config for using the default `Dockerfile` in the repo.
2. Log from the deploy of the subsequently built image:
``` 
id: cannot find name for user ID 1000140000
mkdir: cannot create directory '../sonatype-work/nexus3': Permission denied
mkdir: cannot create directory '../sonatype-work/nexus3': Permission denied
Java HotSpot(TM) 64-Bit Server VM warning: Cannot open file ../sonatype-work/nexus3/log/jvm.log due to Permission denied

Warning:  Cannot open log file: ../sonatype-work/nexus3/log/jvm.log
Warning:  Forcing option -XX:LogFile=/tmp/jvm.log
Unable to update instance pid: Unable to create directory /nexus-data/instances
/nexus-data/log/karaf.log (Permission denied)
Unable to update instance pid: Unable to create directory /nexus-data/instances
```
3. Run https://github.com/sonatype/docker-rhel-nexus/blob/master/OpenShift/init.sh#L47
4. Redeploy
5. Things work.



**RHEL Dockerfile**

1. Change the existing build config  created from the previous test to point to `Dockerfile.rhel7` and delete the generated lines that specify the build is `FROM centos:centos`. This results in the expected build `FROM registry.access.redhat.com/rhel7/rhel`
2. The log from the deploy of this image has the same issue the ticket was open for:
``` 
No suitable Java Virtual Machine could be found on your system.
The version of the JVM must be at least 1.8 and at most 1.8.
Please define INSTALL4J_JAVA_HOME to point to a suitable JVM.
```
3. Run https://github.com/sonatype/docker-rhel-nexus/blob/master/OpenShift/init.sh#L47
4. Redeploy
5. Things still fail.

**Comment by ghost** [comment]
Hi, @sherl0cks.

It is not the same using CentOS or using RHEL as a base image. RHEL uses subscription-manager to enable/disable repos, but when you use RHEL as a docker image, the host that builds the image is the one that should be subscribed.

All these details on https://access.redhat.com/solutions/1443553

Other than that, I would like to add something about the "new-app" command. You can use a template using the new-app. In your case, just use:

```
oc new-app -f https://raw.githubusercontent.com/sonatype/docker-rhel-nexus/master/OpenShift/nexus.json
```

and this is exactly one of the steps in the init.sh script. But also it does several other things, like granting permissions and instantiating the template.

**Comment by sherl0cks** [comment]
Ramon,

The subject of a host with subscriptions was brought up earlier in this
thread. I assure you my build host is properly subscribed. Again the issue
is on deploy, not during s2i build, so subscriptions aren't the issue here.
Let's focus on the specific issue described related to the JVM at runtime
please.

Regarding the template, we have two problems

#1 I dont want to use a template
#2 The template you have referenced in this repo uses a build config with a
docker strategy that does not specify the docker path; therefore, s2i will
default to use the CentOS Dockerfile. As I've described, I want to use a
RHEL baseimage, not CentOS.

Right now, I'm not convinced that anyone else is actually testing the
Dockerfile.rhel7 . If you are, I'd like step by step instructions to
reproduce. Perhaps the confusion here is that we have two Dockerfiles in
the same repo?

**Comment by ghost** [comment]
Hi, @sherl0cks.

I have confirmed the template was using centos instead of rhel for base image. I have requested a PR to split the template in two, one for rhel and one for centos, so anyone on Origin or in OCP can use the corresponding one.

I have tested both, in your case with RHEL as a base image, steps to make it work:

- Create a pv at least with RWO capabilities.
- For easy testing, I have created a new project called ci.
- The policy grant should be done with an user with enough privileges (cluster-admin)
- Until the PR merge, you should use the forked repo (gordillo-ramon-redhat). This template references the Dockerfile in the original repo, so no Dockerfile is modified.
```
oc adm new-project ci
oc adm policy add-scc-to-group anyuid system:serviceaccounts:ci
oc create -f https://raw.githubusercontent.com/gordillo-ramon-redhat/docker-rhel-nexus/master/OpenShift/nexus-rhel.json -n ci
oc start-build nexus -n ci  --wait
oc new-app --template=ci/nexus -n ci
```

**Comment by whyjustin** [comment]
> Until the PR merge, you should use the forked repo (gordillo-ramon-redhat). This template references the Dockerfile in the original repo, so no Dockerfile is modified.

PR Merged via https://github.com/sonatype/docker-rhel-nexus/commit/339e3d119ea98eae0887f94688c8131a75d7879e

**Comment by whyjustin** [comment]
This seems to be fixed via 339e3d1
